### DISCUSSION TITLE
On the stm32h7b3i-dk port, how do I use the spi and gpio pins?

### DISCUSSION THREAD

**Discussion by 4riceotter** [micropython/micropython#11770] [STM32 / Pyboard]
Hi, All,

    With @jimmo and @dpgeorge 's help, 

     https://github.com/orgs/micropython/discussions/11584

I was able to to get started with micropython on the stm32h7b3i-dk board.  Borrowing from some examples from the nucleo-stm32h743i blog,

    https://blog.csdn.net/weixin_39861920/article/details/110035700?

I got the following working to blink the two LED's on board with a timer:
```
tm = pyb.Timer(6, freq=2)
tm.callback(lambda t: pyb.LED(1).toggle())
tm2 = pyb.Timer(7,freq=5)
tm2.callback(lambda t: pyb.LED(2).toggle())
```
or the User switch:
```
sw=pyb.Switch()
sw()
while 1:
    if sw():
        pyb.LED(2).on()
    else:
        pyb.LED(2).off()
```
![blinky](https://github.com/micropython/micropython/assets/101532401/37d63e05-e9de-4739-9be3-b54e2a2cd019)

     
But how do I use the on-board GPIO and SPI pins?  With my previous micropython experience on the Raspberry Pi Pico, the micropython doc tells me which pins I can use for the SPI ports, and all available GPIO pins are addressable by their names.  I tried a few guesses on the GPIO pin names, but they returned with errors.  Any ideas?  

    Your suggestions and advice will be greatly appreciated.  Thanks in advance!

Benjamin

**Comment by jimmo** [comment]
@4riceotter What have you tried so far?

https://docs.micropython.org/en/latest/library/machine.Pin.html
https://docs.micropython.org/en/latest/library/machine.SPI.html

On stm32, you can specify pins either by their CPU name (e.g. `"A6"` _(Edit: sorry, wrote PA6 earlier)_) or their board name (depends on the board, not many defined for this particular board, see https://github.com/micropython/micropython/blob/master/ports/stm32/boards/STM32H7B3I_DK/pins.csv).

**Reply by 4riceotter** [reply]
@jimmo Thanks for the reply and pointers.
Coming from my good experience from micropython on the Pi Pico, I have naively assumed all port pins on connectors or LEDs on the stm32h7b3i-dk board are addressable, but: 

>>> from machine import Pin
>>> pin1 = Pin('PG2',Pin.OUT)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: Pin(PG2) doesn't exist

>>> pin1 = Pin('PD3',Pin.OUT)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: Pin(PD3) doesn't exist

and, from the doc for SPI for pyboard port,

>>> from pyb import SPI
>>> spi1 = SPI(1, SPI.CONTROLLER, baudrate=200000, polarity=1, phase=0)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: SPI(1) doesn't exist  

I figured that my luck is not enough to take me to Las Vegas, and it's time to call for help instead. I assume somewhere there should be a file to show what port pins are addressable as GPIO pins.  Even when I tried those pin names on the pins.csv list, I still got an error (except G2 & G11)!  Aah -- so I should just drop the letter 'P' in the port pin name for use in the Pin function?

OK, that solves the GPIO question (I assume).  The question that still remains: what name and pins should I use in the SPI functions?  Again, greatest appreciation to you for your help!

Benjamin

**Reply by 4riceotter** [reply]
![image](https://github.com/micropython/micropython/assets/101532401/683b21c4-e8a2-4066-bf53-6a2467a9e657)
This is what my attempts should look like.

**Reply by jimmo** [reply]
Sorry yes that was a typo, I should have written A6 not PA6.

FYI you can list the known pins using:

```
>>> dir(machine.Pin.cpu)
['__class__', '__name__', '__bases__', '__dict__', 'A0', 'A1', 'A10', 'A11', 'A12', 'A13', 'A14', 'A15', 'A2', 'A3', 'A4', 'A5', 'A6', 'A7', 'A8', 'A9', 'B0', 'B1', 'B10', 'B11', 'B12', 'B13', 'B14', 'B15', 'B2', 'B3', 'B4', 'B5', 'B6', 'B7', 'B8', 'B9', 'C0', 'C1', 'C10', 'C11', 'C12', 'C13', 'C2', 'C3', 'C4', 'C5', 'C6', 'C7', 'C8', 'C9', 'D2']
>>> dir(machine.Pin.board)
['__class__', '__name__', '__bases__', '__dict__', 'LED_BLUE', 'LED_GREEN', 'LED_RED', 'LED_YELLOW', 'MMA_AVDD', 'MMA_INT', 'SD', 'SD_CK', 'SD_CMD', 'SD_D0', 'SD_D1', 'SD_D2', 'SD_D3', 'SD_SW', 'SW', 'USB_DM', 'USB_DP', 'USB_ID', 'USB_VBUS', 'X1', 'X10', 'X11', 'X12', 'X17', 'X18', 'X19', 'X2', 'X20', 'X21', 'X22', 'X3', 'X4', 'X5', 'X6', 'X7', 'X8', 'X9', 'Y1', 'Y10', 'Y11', 'Y12', 'Y2', 'Y3', 'Y4', 'Y5', 'Y6', 'Y7', 'Y8', 'Y9']
```

Any of those results can be passed to machine.Pin (i.e. `machine.Pin("X11")`, or you can use `machine.Pin.cpu.A6` or `machine.Pin.board.X11` to access the pin object directly (this is my preferred way, but only supported currently on stm32/rp2).

Looking at the SPI... it looks like this board definition is missing the configuration of the SPI peripherals... (argh). If you want to make this work, I suggest looking at e.g. https://github.com/micropython/micropython/blob/master/ports/stm32/boards/PYBD_SF2/mpconfigboard.h#L131
You don't need to define the `NAME`, only the default pins. Once that's done then the SPI peripheral with that ID will be available for `machine.SPI` / `pyb.SPI`  (unless you specifically need the latter, use the former).

**Reply by robert-hh** [reply]
> but only supported currently on stm32/rp2

As well on mimxrt and samd.

**Reply by 4riceotter** [reply]
@jimmo Great, I'll try that tomorrow.
BTW, I did not know about the dir() function. Thanks for pointing that out!

Benjamin

**Reply by 4riceotter** [reply]
Hi,

    I have been able to clone the micropython project and successfully compiled with BOARD=STM32H7B3I_DK.  REPL works, so I proceeded with adding the lines for spi in the file  mpconfigboard.h, using the pins on the STMod+ connector:

#define MICROPY_HW_SPI2_NSS     (pin_A11)
#define MICROPY_HW_SPI2_SCK     (pin_A12)
#define MICROPY_HW_SPI2_MISO    (pin_C2)
#define MICROPY_HW_SPI2_MOSI    (pin_C3)

Micropython recompiles and REPL still works.  But when I tried to instantiate spi I keep getting errors:

>>> from machine import SPI
>>> sp = SPI(baudrate=1000000, polarity=0, phase=0, bits=8, firstbit=SPI.MSB)
Warning: SPI(-1, ...) is deprecated, use SoftSPI(...) instead
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: must specify all of sck/mosi/miso

>>> sp = SPI(2, baudrate=1000000, polarity=0, phase=0, bits=8, firstbit=SPI.MSB)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: SPI(2) doesn't exist

regardless what value I used for the 1st argument.  The example from  pyb.SPI also failed.  How can I tell if SPI is defined correctly, and what argument should I use for the instantiation?

Thanks again for your help!

Benjamin
